zokrates_core/src/zir/mod.rs

```diff
@@ -422,6 +422,39 @@ pub enum BooleanExpression<'ast, T> {
     ),
 }
 
+pub struct ConjunctionIterator<T> {
+    current: Vec<T>,
+}
+
+impl<'ast, T> Iterator for ConjunctionIterator<BooleanExpression<'ast, T>> {
+    type Item = BooleanExpression<'ast, T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.current
+            .pop()
+            .map(|n| match n {
+                BooleanExpression::And(box left, box right) => {
+                    self.current.push(left);
+                    self.current.push(right);
+                    self.next()
+                }
+                n => Some(n),
+            })
+            .flatten()
+    }
+}
+
+impl<'ast, T> IntoIterator for BooleanExpression<'ast, T> {
```

Comment:
I suspect it would be better to have this as part of the BooleanExpression rather than into_iter, it's not immediately clear what `into_iter` means for a tree and we're doing something quite specific here. What about `into_conjunction_iterator`? The only nice thing you get with this std trait is `for in` syntax which isn't a big deal here.

Reply:
yeah good point